### PR TITLE
act: add where block

### PR DESCRIPTION
**Pull request by xwvvvvwx**
Adds support for per act macro substitutions.

Substitution is regex based and occurs as part of `klab build`.  This removes a lot of duplication when writing specs for complex solidity methods (by allowing the `act` spec to shadow internal solidity variables).

An example spec making use of the new `where` block is added [here](https://github.com/dapphub/klab/blob/c7b879345af71040dcbe197d22f8310513478690/examples/where/src/specification.act.md) as part of this PR.

### CODE REVIEW

File: lib/storage.js [hunk 1]
```diff
@@ -1,7 +1,6 @@
 const path          = require("path")
 const marked        = require("marked");
 const Config        = require("../lib/config.js");
-const {parseAct}    = require("../lib/build.js");
```
Comment (xwvvvvwx) [inline lib/storage.js]: unrelated change, just removing an unused import that I stumbled over.

**Comment by xwvvvvwx** [comment]
CI passing here: https://buildbot.dapp.ci/#/builders/2/builds/595

File: acts.md [hunk 2]
```diff
@@ -201,4 +201,13 @@ Lists the behaviours that this spec calls. If the listed behaviour's `_pass` pro
...
 ```
-    
+
+### where
+Allows for the definition of named expressions. All instances of the name on the LHS of the `where` binding will be expanded to the RHS as part of `klab build`. Helps to reduce duplication in specs for methods that have lots of intermediate variables.
+
+```
+where
+
+    Var => X + Y
```
Comment (MrChico) [inline acts.md]: I would use := or just = instead of the rewrite arrow. The equality makes it clear that this expression is really just used as a name for the lobger expression, the rewrite arrow introduces questions like "in what order are things rewritten?"

Comment (asymmetric) [reply acts.md]: @MrChico if this `where` is different from the act2 `where` (e.g. cause act2 `where` is recursive), it could be smart to not use the same symbol.

Could it even be a feature to provide recursive and non-recursive assignments/substitutions?

Not sure if you're thinking of completely breaking backwards compat with act2.

File: lib/build.js [hunk 3]
```diff
@@ -979,6 +1010,27 @@ const buildActs = (config, act_proofs) => {
...
+      from: b.from,
+      to: b.to.replace(bindings[i].from, bindings[i].to)
+    }))
+  }
+  return bindings
+}
+
+// apply where bindings to the spec
+const applyWhereBindings = (spec, bindings) => {
+  if (typeof bindings === "undefined") return spec
```
Comment (asymmetric) [inline lib/build.js]: Nit: I think it would be nice to move this check to an `extractWhereBindings` function, so that the "meaty" functions could only focus on the actual logic.

File: lib/build.js [hunk 4]
```diff
@@ -979,6 +1010,27 @@ const buildActs = (config, act_proofs) => {
 
 }
 
+// apply every where binding in turn to every other where binding
+const expandWhereBindings = (bindings) => {
+  for (i = 0; i < bindings.length; i++) {
```
Comment (asymmetric) [inline lib/build.js]: Have you thought of using an inner loop instead, rather than mixing `for` and `map` for iteration? I think it could be slightly more clear.

Would probably be more efficient too (since you can just modify the `to` key in place rather than constructing a new array), but that's a minor concern in this case I guess.

```js
const expandWhereBindings = (bindings) => {
  for (i = 0; i < bindings.length; i++) {
    for (j = 0; j < bindings.length; j++) {
      bindings[j].to = bindings[j].to.replace(bindings[i].from, bindings[i].to);
    }
  }
  return bindings
}
```

I'm fine either way btw, it's mostly aesthetics.

File: acts.md (hunk 2 above)
Comment (MrChico) [reply acts.md]: This has less to do with act2 compat and more to do with aesthetics. To me its cleaner to define the rhs as an expression rather than a term which can be further rewritten

Comment (MrChico) [reply acts.md]: This has less to do with act2 compat and more to do with aesthetics. To me its cleaner to define the rhs as an expression rather than a term which can be further rewritten

Comment (MrChico) [reply acts.md]: Obviously only a syntactic concern, I just find it cleaner

Comment (MrChico) [reply acts.md]: Act behaviours should define the atomic steps in a transition system. Each behaviour defines one (large step) rewrite. The `where` section can be used to help expressing the rhs of this rewrite, but introducing helpers as rewrite rules themselves is too nuch rewriting maximalism for my tastes

Comment (xwvvvvwx) [reply acts.md]: Not 100% sure what you mean with: `an expression rather than a term which can be further rewritten`, but the current implementation allows the RHS of macros in the `where` block to be rewritten by other macros in the `where` block. Do you see this as a problem?

Comment (MrChico) [reply acts.md]: I don't see that as a problem, but I think thinking of what the `where` header does as defining further rewrites is not a perspective that I would like to proliferate

Comment (xwvvvvwx) [reply acts.md]: ah ok, you want to disambiguate them from `K` rewrite rules. Yes, this makes sense.

File: examples/where/src/prelude.smt2.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+```smt2
+(set-option :auto-config false)
+(set-option :smt.mbqi false)
+```
```
[the remark this answers is not in the document]
Comment (xwvvvvwx) [reply examples/where/src/prelude.smt2.md]: I'm not 100% sure, but afaict they are only required if you are using the `patterns` feature (see the `patterns` section in the [z3 guide](https://rise4fun.com/Z3/tutorial/guide)). Removed. Thanks for asking :+1: :pray:

File: lib/build.js (hunk 4 above)
Comment (xwvvvvwx) [reply lib/build.js]: Agree this is tidier. Thanks :+1: :pray:

File: acts.md (hunk 2 above)
Comment (xwvvvvwx) [reply acts.md]: done.

File: lib/build.js (hunk 3 above)
Comment (xwvvvvwx) [reply lib/build.js]: done.
